### ISSUE TITLE
Add Dune to the Coq Platform

### ISSUE THREAD

**Issue by palmskog** [coq/platform#11]
[Dune](https://dune.build) is a composable and opinionated build system for Coq and OCaml. 

Dune is a robust alternative to fragile Makefiles and using `coq_makefile` for building Coq projects, and nowadays even supports extraction to OCaml. Dune support for Coq is currently at language version 0.2 and is not considered final, but I use Dune in CI for all my projects and it works very well. Long-term stable Dune support for Coq is expected in the near future.

I propose that the Coq Platform includes the latest stable Dune release, which at the time of writing is version 2.6.2.

The Coq maintainer for Dune is @ejgallego.

**Comment by ejgallego** [comment]
Small note , @rgrinberg is also a maintainer. Indeed dune is already needed by the OCaml tooling / base images I think.

**Comment by gares** [comment]
I'm not so sure dune qualifies as a "package for the platform".
The current scrips pulls it in since some other packages need it in order to build, which is OK to me.

See the current list of packages for Coq 8.12: https://github.com/coq/platform/blob/master/coq_platform_packages.sh , they are all coq-* packages. Some external dependencies are sometimes pinned to a specific version, like the elpi version compiling fine on windows, but I don't think we need anything that for dune.

FYI, this is the list of packages installed for 8.13, as you can see all the dependencies are in, including dune (2.7), and bison, boost, gmp, perl, ...:
```
The following actions will be performed:
  ∗ install seq                      base         [required by re]
  ∗ install conf-g++                 1.0          [required by coq-interval, coq-flocq, coq-coquelicot, etc.]
  ∗ install camlp5                   7.13         [required by elpi]
  ∗ install conf-findutils           1            [required by coq, coqide]
  ∗ install conf-m4                  1            [required by ocamlfind]
  ∗ install ocaml-secondary-compiler 4.08.1-1     [required by ocamlfind-secondary]
  ∗ install ocamlbuild               0.14.0       [required by coq-quickchick]
  ∗ install conf-bison               1            [required by gappa]
  ∗ install conf-boost               1            [required by gappa]
  ∗ install conf-gmp                 3            [required by gappa]
  ∗ install conf-perl                1            [required by zarith]
  ∗ install conf-pkg-config          1.3          [required by conf-adwaita-icon-theme, conf-gtk3]
  ∗ install conf-which               1            [required by conf-autoconf]
  ∗ install conf-flex                1            [required by gappa]
  ∗ install ocamlfind                1.8.1        [required by coq-quickchick, coq, coq-gappa, coqide]
  ∗ install conf-mpfr                2            [required by gappa]
  ∗ install conf-gtksourceview3      0+2          [required by lablgtk3-sourceview3]
  ∗ install conf-gtk3                18           [required by lablgtk3]
  ∗ install conf-cairo               1            [required by cairo2]
  ∗ install conf-adwaita-icon-theme  1            [required by coqide]
  ∗ install conf-autoconf            0.1          [required by coq-gappa, gappa]
  ∗ install zarith                   1.11         [required by coq]
  ∗ install ocamlfind-secondary      1.8.1        [required by dune]
  ∗ install num                      1.4          [required by coq]
  ∗ install conf-automake            1            [required by gappa]
  ∗ install dune                     2.7.1        [required by lablgtk3, elpi, menhir]
  ∗ install coq                      8.13+beta1
  ∗ install gappa                    1.3.5
  ∗ install result                   1.5          [required by dune-configurator]
  ∗ install re                       1.9.0        [required by elpi]
  ∗ install menhirSdk                20200624     [required by menhir]
  ∗ install menhirLib                20200624     [required by menhir]
  ∗ install coq-unicoq               1.5+8.13
  ∗ install coq-menhirlib            20200624
  ∗ install coq-mathcomp-ssreflect   1.12.0
  ∗ install coq-flocq                3.3.1
  ∗ install coq-ext-lib              0.11.3
  ∗ install coq-equations            1.2.3+8.13
  ∗ install coq-bignums              8.13.0~beta1
  ∗ install coq-aac-tactics          8.13.0
  ∗ install csexp                    1.3.2        [required by dune-configurator]
  ∗ install elpi                     1.12.0
  ∗ install menhir                   20200624
  ∗ install coq-mtac2                1.4+8.13
  ∗ install coq-mathcomp-fingroup    1.12.0
  ∗ install coq-mathcomp-bigenough   1.0.0
  ∗ install coq-coquelicot           3.1.0
  ∗ install coq-gappa                1.4.6
  ∗ install coq-simple-io            1.5.0
  ∗ install dune-configurator        2.7.1        [required by cairo2]
  ∗ install coq-elpi                 1.8.1
  ∗ install coq-compcert             3.8
  ∗ install coq-mathcomp-algebra     1.12.0
  ∗ install coq-mathcomp-finmap      1.5.1
  ∗ install coq-interval             4.1.0
  ∗ install coq-quickchick           1.5.0
  ∗ install cairo2                   0.6.2        [required by lablgtk3]
  ∗ install coq-hierarchy-builder    1.0.0
  ∗ install coq-vst                  2.7
  ∗ install coq-mathcomp-solvable    1.12.0
  ∗ install lablgtk3                 3.1.1
  ∗ install coq-mathcomp-field       1.12.0
  ∗ install lablgtk3-sourceview3     3.1.1        [required by coqide]
  ∗ install coq-mathcomp-real-closed 1.1.2
  ∗ install coq-mathcomp-character   1.12.0
  ∗ install coqide                   8.13+beta1
```
I don't think we should list them all in the platform.

**Comment by Zimmi48** [comment]
I think Dune qualifies for direct inclusion in the platform, and not just as an external dependency, because it is a useful tool to have for any Coq user (or at least it will be when the Coq lang in Dune reaches 1.0). As a bonus, I think it is very much compatible with Windows, but we should verify that using Dune installed using the Windows installer makes it possible to compile a pure Coq project (in particular, without needing OCaml or findlib).

**Comment by palmskog** [comment]
What I figured about including Dune in the platform is that there would be a selected Dune version for every Platform release, confirmed to be compatible with at least the projects in the Platform that use Dune, but ideally many more as well. If it's only an external dependency, you could get any newly released version, regardless if it has regressions.

**Comment by gares** [comment]
> it possible to compile a pure Coq project (in particular, without needing OCaml or findlib).

Right now ocamlc is not in the installer. But I think this is unrelated to dune being "part of the platform".

>  there would be a selected Dune version for every Platform release

I'm OK for selecting one version, but then I guess we need to split the list PACKAGES (which I planned to use also for CI) into COQ_PACKAGES and EXTRA. I don't think the Coq platform should do anything special for CI w.r.t. elpi, gtk, or dune (other than installing the pinned version, I mean). COQ_PACKAGES are linked to Coq's CI, in some way or another, and are the things that are likely to break.

**Comment by gares** [comment]
I mean, it's like the docker image we use for Coq. Dune is in there, and we pick a version we like, but it is not a project in ci-basic-overlays.sh

**Comment by Zimmi48** [comment]
> we should verify that using Dune installed using the Windows installer makes it possible to compile a pure Coq project (in particular, without needing OCaml or findlib).

Actually, last time I checked (https://github.com/ocaml/dune/issues/2042), Dune was not able to properly function without OCaml in PATH (even for a pure Coq project). In this case, it's not really useful to have Dune shipped with the Windows installer if OCaml is not.

**Comment by ejgallego** [comment]
Regarding the version here, I guess that the minimal dune version in the platform will be determined more by the set of OCaml packages installed as what their deps are. Note that dune can be easily vendored tho, so it is not a big deal to have different versions of Dune living together.

[In fact that's how I first proposed to add Dune support to Coq]

**Comment by MSoegtropIMC** [comment]
I agree that dune should be included with an explicit version and also appear in the advertised package lists. We have a few other packages which are not coq packages (elpi and gappa) explicitly included.

Currently I have:
```
dune                     2.9.1       Fast, portable, and opinionated build system
dune-configurator        2.9.1       Helper library for gathering system configuration
```
Is this an appropriate version for both 8.13 and 8.14?

**Comment by ejgallego** [comment]
@MSoegtropIMC yes, 2.9.1 is the choice IMHO.

**Comment by MSoegtropIMC** [comment]
Thanks for the confirmation!

I added an explicit reference to dune.2.9.1 in #143.

Do you think it should be an explicitly selectable package in the Windows installer. The Windows installer inherits opam's dependencies and automatically selects or deselects packages as needed, but packages can be user visible or hidden. Not sure in which category dune should go.

**Comment by MSoegtropIMC** [comment]
For the records: this request misses a maintainer agreement, but since Coq anyway relies on dune now, I guess Coq users anyway rely on that dune is properly maintained in the future, so we can skip that.

**Comment by gares** [comment]
given https://github.com/coq/platform/issues/11#issuecomment-746838690 it should not even be shipped in the win installer

**Comment by ejgallego** [comment]
> given [#11 (comment)](https://github.com/coq/platform/issues/11#issuecomment-746838690) it should not even be shipped in the win installer

But we are shipping OCaml , aren't we; otherwise is native disabled in windows?

**Comment by gares** [comment]
If you use the pkatform scripts you will get dune and ocaml, since coq depends on them. The scripts can pick a good version of those. 

But the binary installer is a trimmed down environment. IIRC it does not include ocaml, nor bash & co. so it should not include dune IMO.

**Comment by MSoegtropIMC** [comment]
@gares : wasn't it stated once that dune is the way to get around the dependency on bash? I would like to include it in the installer so that people can build their own Coq projects with dune - I take it that this is supported meanwhile.

**Comment by Zimmi48** [comment]
Did you see the comment that @gares referred to? Dune fails if OCaml is not in path (this is still true with Dune 2.9), even when compiling a pure Coq project:

```
Error: Program ocamlc not found in the tree or in PATH
```

Therefore, there is no point in including Dune in the installer until this issue is resolved, because it cannot be used to build Coq projects if OCaml is not also shipped in the installer.

**Comment by gares** [comment]
@MSoegtropIMC you know the issues of the ocaml toolchain on windows better than me: if the ocamlc which is in the opam root works even without a cygwin environment then it should be possible to put it in the installer and that would "unblock" dune.

If we are not there yet, then dune cannot be shipped in the installer.

**Comment by MSoegtropIMC** [comment]
@gares : I was talking about using dune for building Coq projects, not OCaml projects.

Also so far the main issue with OCaml on Windows is not OCaml as such, but that most suitable build systems require a shell, which is quite messy to get on Windows.

Btw.: as mentioned a few times I do have C code which implements a reasonable subset of a posix shell - it supports all sorts of redirection (including exotic multi handle stuff), piping an chaining, but no if, case, for, function ... Using this on Windows as OCaml system command would solve some issues, but up to now I always heard that this is not required with dune.

So in short, the question is: is dune the (w)holy grail of system independent build systems or not?

**Comment by gares** [comment]
> even when compiling a pure Coq project:
> 
> ```
> Error: Program ocamlc not found in the tree or in PATH
> ```

I know it makes little sense, but this is how it is. So, no, there's no such a thing.

**Comment by gares** [comment]
FTR, I'm in favor of shipping dune as part of the platform, and including it in the window installer *when* it will work.

**Comment by Zimmi48** [comment]
@MSoegtropIMC I don't see how you interpret this as controversial. We just told you that Dune needs OCaml. If you think that the main issue of OCaml on Windows is the build systems it comes with, then fine, ship both OCaml and Dune. But you cannot ship Dune without OCaml and except it will work to build Coq project (for the time being, at least).

**Comment by MSoegtropIMC** [comment]
@Zimmi48 : with "controversial" I mean there is some sort of discussion which keeps me from closing this issue. I didn't know that dune needs OCaml for compiling Coq projects - I expected it is a compiled binary which reads in files describing the project structure, decides what to do in what order and then calls other compiled binaries. May I ask for what purpose it needs OCaml when compiling Coq code? Or did I get it wrong that dune can handle Coq projects, so that it is only useful for compiling OCaml code?

**Comment by Zimmi48** [comment]
>May I ask for what purpose it needs OCaml when compiling Coq code?

I don't know, but probably for some trivial reason. I only reported this just now to the Dune bug tracker: https://github.com/ocaml/dune/issues/4949

>Or did I get it wrong that dune can handle Coq projects, so that it is only useful for compiling OCaml code?

No, you didn't get that part wrong. It's useful for pure Coq projects.

**Comment by MSoegtropIMC** [comment]
OK, so I guess the conclusion is then to not explicitly include dune (just as dependency) until we find a way to support it on Windows.

I see that as piece of the greater "OCaml and Opam" on Windows story. I think if a few experts sit together for a day or two, we could sort out most of the issues with opam 2.1, so that the remaining work is opam repo maintenance. In the light of the discontinuation of the MinGW Opam project in August, we should try to get this sorted out soon. I am usually available with a week ahead notice (would take vacation for that).

**Comment by ejgallego** [comment]
Umm, so the windows platform installer cannot use `coq_makefile` either?

**Comment by gares** [comment]
If you have a cygwin (eg make and bash), then you can use coq_makefile to compile pure .v projects.
It has been like that since I can remember.

**Comment by gares** [comment]
Technically, coq_makefile runs fine. You need make and bash to run the generated makefile...

**Comment by MSoegtropIMC** [comment]
> Umm, so the windows platform installer cannot use coq_makefile either?

Yes. I did package in the very beginning gnu make but no shell. This did require some minor changes to the make file (say replace rm with del) but people didn't like it because it did hide other make binaries people might have.

So on Windows an installed Coq Platform doesn't come with make and is in the end more for homework. The cygwin "from the sources" installation of course works.

I still think it wouldn't be that hard to make the coq makefiles platform independent - it would be setting variables for things like rm=del, cp=copy, mv=move, and a few other things.

**Comment by MSoegtropIMC** [comment]
FTR: I will remove the explicit reference for dune. This means people get as dependency the latest version of dune if they compile from sources and no dune in installers. I hope we can improve on this soon.

**Comment by MSoegtropIMC** [comment]
Removed with PR (https://github.com/coq/platform/pull/144).

**Comment by ejgallego** [comment]
I'm not sure I follow, dune and ocaml don't need a shell, they run fine in native win32 [the fact that dune requires the ocaml compiler is indeed unfortunately and likely to be solved soon, as it is also a condition for the composed OCaml builds]

**Comment by MSoegtropIMC** [comment]
@ejgallego : the point is that in a "from sources" installation it will anyway be there and for "installed" installations it would requires quite a bit of work on the OCaml side to be usable. So let's work on make it usable and add it back then.
